crates/ruff/src/message/text.rs

```diff
@@ -245,36 +246,40 @@ impl Display for MessageCodeFrame<'_> {
 }
 
 fn replace_whitespace(source: &str, annotation_range: TextRange) -> SourceCode {
-    static TAB_SIZE: TextSize = TextSize::new(4);
+    static TAB_SIZE: u32 = 4; // TODO(jonathan): use `pycodestyle.tab-size`
 
     let mut result = String::new();
     let mut last_end = 0;
     let mut range = annotation_range;
     let mut column = 0;
 
-    for (index, m) in source.match_indices(['\t', '\n', '\r']) {
-        match m {
-            "\t" => {
-                let tab_width = TAB_SIZE - TextSize::new(column % 4);
+    for (index, c) in source.chars().enumerate() {
+        match c {
+            '\t' => {
+                let tab_width = TAB_SIZE - column % TAB_SIZE;
+                column += tab_width;
 
                 if index < usize::from(annotation_range.start()) {
-                    range += tab_width - TextSize::new(1);
+                    range += TextSize::new(tab_width - 1);
                 } else if index < usize::from(annotation_range.end()) {
-                    range = range.add_end(tab_width - TextSize::new(1));
+                    range = range.add_end(TextSize::new(tab_width - 1));
                 }
 
                 result.push_str(&source[last_end..index]);
 
-                for _ in 0..u32::from(tab_width) {
+                for _ in 0..tab_width {
                     result.push(' ');
                 }
 
                 last_end = index + 1;
             }
-            "\n" | "\r" => {
+            '\n' | '\r' => {
                 column = 0;
             }
-            _ => unreachable!(),
+            #[allow(clippy::cast_possible_truncation)]
+            _ => {
+                column += c.width().unwrap_or(0) as u32;
+            }
```

Comment:
Lol really, did I forget to increment the column? That's embarrassing. 


My understanding is that tabs don't operate on character width but on character counts, meaning, we should increment the column by one for every character (except newlines, because they reset the column to 0)

```suggestion
		match c {
			'\t' => {
                let tab_width = TAB_SIZE - column % TAB_SIZE;

                if index < usize::from(annotation_range.start()) {
                    range += TextSize::new(tab_width - 1);
                } else if index < usize::from(annotation_range.end()) {
                    range = range.add_end(TextSize::new(tab_width - 1));
                }

                result.push_str(&source[last_end..index]);

                for _ in 0..tab_width {
                    result.push(' ');
                }

                last_end = index + 1;
            }
            '\n' | '\r' => {
                column = 0;
                continue;
            },
            _ => {}
		}
		
		column += 1;
```